File: backend/app/schemas/ai.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field

from app.schemas.dashboard import DashboardWidget


class ChartHint(BaseModel):
    type: Literal["line", "bar", "pie", "none"] = "none"
    x_column: str | None = None
    y_column: str | None = None


class ChartBuildSpec(BaseModel):
    chart_type: str
    x_column: str | None = None
    y_column: str | None = None
    aggregation: str = "sum"


class ChatRequest(BaseModel):
    dataset_id: str
    session_id: str | None = None
    message: str = Field(min_length=1, max_length=4000)


class ChatResponse(BaseModel):
    session_id: str
    answer: str
    sql: str | None = None
    result: list[dict] | None = None
    chart_hint: ChartHint | None = None
    charts: list[DashboardWidget] | None = None
    tokens_used: int = 0


class ChatSessionItem(BaseModel):
    id: str
    dataset_id: str
    created_at: datetime
    updated_at: datetime
    message_count: int = 0


class ChatMessageItem(BaseModel):
    id: str
    role: str
    content: str
    sql_generated: str | None = None
    result_json: list | dict | None = None
    chart_hint: ChartHint | None = None
    charts: list[DashboardWidget] | None = None
    tokens_used: int | None = None
    created_at: datetime


class ApiResponse(BaseModel):
    success: bool = True
    data: dict | list | None = None
    error: dict | None = None

File: frontend/src/types/ai.ts
export interface ChartHint {
  type: "line" | "bar" | "pie" | "none";
  x_column?: string | null;
  y_column?: string | null;
}

export interface ChatChartWidget {
  id: string;
  type: string;
  title: string;
  config: Record<string, string | null>;
  data: Record<string, unknown> | Array<Record<string, string | number>>;
}

export interface ChatResponse {
  session_id: string;
  answer: string;
  sql?: string | null;
  result?: Record<string, unknown>[] | null;
  chart_hint?: ChartHint | null;
  charts?: ChatChartWidget[] | null;
  suggestions?: string[] | null;
  tokens_used: number;
}

export interface ChatSessionItem {
  id: string;
  dataset_id: string;
  created_at: string;
  updated_at: string;
  message_count: number;
}

export interface ChatMessageItem {
  id: string;
  role: "user" | "assistant" | "system";
  content: string;
  sql_generated?: string | null;
  result_json?: Record<string, unknown>[] | null;
  chart_hint?: ChartHint | null;
  charts?: ChatChartWidget[] | null;
  suggestions?: string[] | null;
  tokens_used?: number | null;
  created_at: string;
}

export interface ChatRequest {
  dataset_id: string;
  session_id?: string | null;
  message: string;
}
